File: service/jira_service.py
import os
from dotenv import load_dotenv
from jira import JIRA
from jira.exceptions import JIRAError
from requests import Session

# Load environment variables from service/.env
ENV_PATH = os.path.join(os.path.dirname(__file__), ".env")
load_dotenv(ENV_PATH)

_JIRA_CLIENT = None

def connect_jira():
    """
    Returns a cached JIRA client after validating env vars and server connectivity.
    Raises RuntimeError on missing config or connection/auth failures.
    """
    global _JIRA_CLIENT
    if _JIRA_CLIENT:
        return _JIRA_CLIENT

    base = os.getenv("JIRA_URL")
    email = os.getenv("JIRA_EMAIL")
    token = os.getenv("JIRA_API_TOKEN")

    if not base or not email or not token:
        raise RuntimeError("JIRA_URL, JIRA_EMAIL and JIRA_API_TOKEN must be set in service/.env")

    options = {"server": base}
    try:
        session = Session()
        _JIRA_CLIENT = JIRA(options=options, basic_auth=(email, token), session=session)
        # connectivity/auth check
        info = _JIRA_CLIENT.server_info()
        print("Connected to Jira. Server version:", info.get("version"))
        return _JIRA_CLIENT
    except JIRAError as e:
        raise RuntimeError(f"Failed to connect to Jira: {e}") from e

def get_backlog_issues(jql="assignee=currentUser()"):
    """
    Returns search results for the provided JQL.
    """
    jira = connect_jira()
    try:
        return jira.search_issues(jql)
    except JIRAError as e:
        raise RuntimeError(f"Error searching issues: {e}") from e

def create_issue(project_key, summary, description, issue_type="Task"):
    """
    Creates an issue after validating that the issue type exists in the project.
    Returns the created issue object.
    """
    jira = connect_jira()
    try:
        meta = jira.createmeta(projectKeys=project_key)
        types = [t["name"] for t in meta["projects"][0]["issuetypes"]]
    except Exception:
        raise RuntimeError(f"Unable to read project metadata for {project_key}")

    if issue_type not in types:
        raise RuntimeError(f"Issue type '{issue_type}' not available in project {project_key}. Available: {types}")

    issue_dict = {
        "project": {"key": project_key},
        "summary": summary,
        "description": description,
        "issuetype": {"name": issue_type},
    }
    try:
        return jira.create_issue(fields=issue_dict)
    except JIRAError as e:
        raise RuntimeError(f"Failed to create issue: {e}") from e

def update_issue(issue_key, fields):
    """
    Updates fields of an existing issue and returns the refreshed issue object.
    """
    jira = connect_jira()
    try:
        issue = jira.issue(issue_key)
        issue.update(fields=fields)
        return jira.issue(issue_key)
    except JIRAError as e:
        raise RuntimeError(f"Failed to update issue {issue_key}: {e}") from e

def transition_issue(issue_key, transition_name):
    """
    Attempts to transition an issue by name. Returns a message indicating success
    or available transitions when the requested one is not found.
    """
    jira = connect_jira()
    try:
        issue = jira.issue(issue_key)
        transitions = jira.transitions(issue)
        tid = next((t["id"] for t in transitions if t["name"] == transition_name), None)
        if tid:
            jira.transition_issue(issue, tid)
            return f"Issue {issue_key} transitioned to {transition_name}"
        else:
            names = [t["name"] for t in transitions]
            return f"Transition '{transition_name}' not found. Available transitions: {names}"
    except JIRAError as e:
        raise RuntimeError(f"Failed to transition issue {issue_key}: {e}") from e

File: service/jira_service.js
import path from "path";
import { fileURLToPath } from "url";
import dotenv from "dotenv";
import JiraApi from "jira-client";
import axios from "axios";

// -----------------------------------------------------
// Resolve correct .env path for ESM (THIS FIXES YOUR ISSUE)
// -----------------------------------------------------
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

dotenv.config({ path: path.join(__dirname, ".env") });

// -----------------------------------------------------
// GLOBAL JIRA CLIENT HOLDER
// -----------------------------------------------------
let jira = null;

// -----------------------------------------------------
// Connect to Jira
// -----------------------------------------------------
function connectJira() {
  if (jira) return jira;

  const base = process.env.JIRA_URL;
  const email = process.env.JIRA_EMAIL;
  const token = process.env.JIRA_API_TOKEN;

  if (!base || !email || !token) {
    throw new Error("Missing Jira credentials in .env");
  }

  const host = base.replace(/^https?:\/\//, "").replace(/\/$/, "");

  jira = new JiraApi({
    protocol: "https",
    host,
    username: email,
    password: token,
    apiVersion: "2",
    strictSSL: true
  });

  return jira;
}

// -----------------------------------------------------
// GET ISSUES USING NEW JIRA CLOUD ENDPOINT
// -----------------------------------------------------
async function getAssignedIssues(jql = "assignee=currentUser()") {
  const base = process.env.JIRA_URL;
  const email = process.env.JIRA_EMAIL;
  const token = process.env.JIRA_API_TOKEN;

  const auth = Buffer.from(`${email}:${token}`).toString("base64");

  try {
    const res = await axios.post(
      `${base}/rest/api/3/search/jql`,
      { jql },
      {
        headers: {
          Authorization: `Basic ${auth}`,
          Accept: "application/json",
          "Content-Type": "application/json"
        }
      }
    );

    return res.data.issues || [];
  } catch (err) {
    console.error(">>> Jira search error:", err.response?.data || err.message);
    throw new Error(`Error searching issues: ${err.message}`);
  }
}

// -----------------------------------------------------
// CREATE ISSUE
// -----------------------------------------------------
async function createIssue({ projectKey, summary, description, issueType = "Task" }) {
  const client = connectJira();
  if (!projectKey) projectKey = process.env.JIRA_PROJECT;

  const payload = {
    fields: {
      project: { key: projectKey },
      summary,
      description,
      issuetype: { name: issueType }
    }
  };

  try {
    const result = await client.addNewIssue(payload);
    return result;
  } catch (err) {
    throw new Error(`Failed to create issue: ${err.message}`);
  }
}

// -----------------------------------------------------
// UPDATE ISSUE
// -----------------------------------------------------
async function updateIssue(issueKey, fields) {
  const client = connectJira();
  try {
    await client.updateIssue(issueKey, { fields });
    return await client.findIssue(issueKey);
  } catch (err) {
    throw new Error(`Failed to update issue ${issueKey}: ${err.message}`);
  }
}

// -----------------------------------------------------
// TRANSITION ISSUE
// -----------------------------------------------------
async function transitionIssue(issueKey, transitionName) {
  const client = connectJira();
  try {
    const transitions = await client.listTransitions(issueKey);
    const available = transitions.transitions || transitions;

    const match = available.find((t) => t.name === transitionName);

    if (!match) {
      const availableNames = available.map((t) => t.name).join(", ");
      return `Transition '${transitionName}' not found. Available: ${availableNames}`;
    }

    await client.transitionIssue(issueKey, { transition: { id: match.id } });
    return `Issue ${issueKey} transitioned to ${transitionName}`;
  } catch (err) {
    throw new Error(`Failed to transition issue ${issueKey}: ${err.message}`);
  }
}

// -----------------------------------------------------
// EXPORTS
// -----------------------------------------------------
export {
  connectJira,
  getAssignedIssues,
  createIssue,
  updateIssue,
  transitionIssue
};

// -----------------------------------------------------
// SELF-TEST MODE (node jira_service.js --test)
// -----------------------------------------------------
if (process.argv.includes("--test")) {
  console.log(">>> Running Jira service test...");

  try {
    const issues = await getAssignedIssues();
    console.log(`>>> Issues assigned to you: ${issues.length}`);

    const created = await createIssue({
      summary: "Test issue from terminal",
      description: "This issue was created via Node.js Jira integration."
    });

    console.log(">>> Created:", created.key);

    const updated = await updateIssue(created.key, {
      summary: "Updated test issue"
    });

    console.log(">>> Updated:", updated.key);

    const transitioned = await transitionIssue(created.key, "In Progress");
    console.log(">>>", transitioned);

    console.log(">>> All tests finished successfully.");
  } catch (err) {
    console.error(">>> TEST ERROR:", err.message);
  }
}
